witness/internal/witness/witness.go

```diff
@@ -0,0 +1,175 @@
+// Copyright 2021 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package witness
+
+import (
+	"fmt"
+
+	"github.com/google/trillian-examples/formats/log"
+	"github.com/google/trillian/merkle/hashers"
+	"github.com/google/trillian/merkle/logverifier"
+	"golang.org/x/mod/sumdb/note"
+)
+
+type Chkpt struct {
+	Parsed *log.Checkpoint
+	Raw    []byte
+}
+
+type ChkptStorage interface {
+	// GetLatest returns the latest checkpoint for a given log.
+	GetLatest(logPK string) (*Chkpt, error)
+
+	// SetCheckpoint adds a checkpoint to the storage for a given log.
+	SetCheckpoint(logPK string, c Chkpt) error
+}
+
+type Witness struct {
+	db     ChkptStorage
+	Signer note.Signer
+	Hasher hashers.LogHasher
+	LogV   logverifier.LogVerifier
+	SigVs  []note.Verifier
+}
+
+// NewWitness creates a new witness, which initially has no logs to follow.
+func NewWitness(h hashers.LogHasher, db ChkptStorage, sk string) (*Witness, error) {
```

Comment:
Perhaps this is YAGNI, but the hasher ought to be a property of the log. Or we should document that each witness can only witness logs with a pre-determined hash strategy.

Reply:
Okay, witness now maintains a mapping from logIDs to log information, including hasher, so it can now support whatever hashing strategy each log wants.